File: 0208.Implement-Trie(Prefix-Tree)/step2.py
from __future__ import annotations


class Trie:
    class TrieNode:
        def __init__(self):
            self.children: dict[str, Trie.TrieNode] = {}
            self.is_end = False

    def __init__(self) -> None:
        self.root = Trie.TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = Trie.TrieNode()
            node = node.children[c]
        node.is_end = True

    def _search_child(self, sub_word: str) -> Trie.TrieNode | None:
        node = self.root
        for c in sub_word:
            if c not in node.children:
                return None
            node = node.children[c]
        return node

    def search(self, word: str) -> bool:
        node = self._search_child(word)
        return node is not None and node.is_end

    def startsWith(self, prefix: str) -> bool:
        return self._search_child(prefix) is not None


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: 0208.Implement-Trie(Prefix-Tree)/step2.cpp
#include <array>
#include <memory>
#include <string>

class Trie {
    class TrieNode {
    public:
        std::array<std::unique_ptr<TrieNode>, 26> children{};
        bool is_end = false;
    };

    std::unique_ptr<TrieNode> root_;

    const TrieNode* searchChild(const std::string& word) const {
        const TrieNode* node = root_.get();
        for (char c : word) {
            const int i = c - 'a';
            if (!node->children[i]) {
                return nullptr;
            }
            node = node->children[i].get();
        }
        return node;
    }

public:
    Trie() : root_(std::make_unique<TrieNode>()) {}

    void insert(std::string word) {
        TrieNode* node = root_.get();
        for (char c : word) {
            const int i = c - 'a';
            if (!node->children[i]) {
                node->children[i] = std::make_unique<TrieNode>();
            }
            node = node->children[i].get();
        }
        node->is_end = true;
    }

    bool search(std::string word) {
        const TrieNode* node = searchChild(word);
        return node != nullptr && node->is_end;
    }

    bool startsWith(std::string prefix) {
        return searchChild(prefix) != nullptr;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */